contrib/mesos/pkg/minion/tasks/task.go

```diff
@@ -241,61 +251,75 @@ type exitError interface {
 }
 
 func taskRunning(t *Task) taskStateFn {
-	waiter := t.cmd.Wait
-	var sendOnce sync.Once
-	trySend := func(wr *Completion) {
-		// guarded with once because we're only allowed to send a single "result" for each
-		// process termination. for example, if Kill() results in an error because Wait()
-		// already completed we only want to return a single result for the process.
-		sendOnce.Do(func() {
-			t.tryComplete(wr)
-		})
-	}
 	// listen for normal process completion in a goroutine; don't block because we need to listen for shouldQuit
 	waitCh := make(chan *Completion, 1)
 	go func() {
-		wr := &Completion{}
+		wr := &Completion{name: t.name}
 		defer func() {
 			waitCh <- wr
 			close(waitCh)
 		}()
 
-		if err := waiter(); err != nil {
+		if err := t.cmd.Wait(); err != nil {
 			if exitError, ok := err.(exitError); ok {
 				if waitStatus, ok := exitError.Sys().(syscall.WaitStatus); ok {
-					wr.name = t.name
 					wr.code = waitStatus.ExitStatus()
 					return
 				}
 			}
 			wr.err = fmt.Errorf("task wait ended strangely for %q: %v", t.bin, err)
-		} else {
-			wr.name = t.name
 		}
 	}()
 
-	// wait for the process to complete, or else for a "quit" signal on the task (at which point we'll attempt to kill manually)
-	select {
-	case <-t.shouldQuit:
-		// check for tie
+	t.tryComplete(t.awaitDeath(defaultKillGracePeriod, waitCh))
```

Comment:
I think I see what you mean, will give it a shot

On Thu, Sep 24, 2015 at 1:29 PM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/minion/tasks/task.go
> <https://github.com/kubernetes/kubernetes/pull/14489#discussion_r40346702>
> :
>
> >  		}
> >  	}()
> >
> > -	// wait for the process to complete, or else for a "quit" signal on the task (at which point we'll attempt to kill manually)
> > -	select {
> > -	case <-t.shouldQuit:
> > -		// check for tie
> > +	t.tryComplete(t.awaitDeath(defaultKillGracePeriod, waitCh))
>
> Wouldn't it more natural to move the outer select here?
>
> select {
>     case <-t.shouldQuit:
>         return taskShouldQuit(defaultGracePeriod, waitCh)
>     case wr <- waitCh:
>         return taskShouldRestart
> }
>
> Then you can reduce the nesting in taskShouldQuit compared to now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14489/files#r40346702>.
>